homeassistant/components/dreo/coordinator.py

```diff
@@ -0,0 +1,144 @@
+        """Create appropriate device data based on device type."""
+        if (
+            coordinator.device_type == FAN_DEVICE.get("type")
+            and coordinator.device_model
+        ):
+            return FanDataStrategy.process_data(
+                coordinator.device_model, base_data, status
+            )
+        return DreoFanDeviceData(**base_data)
+
+
+class FanDataStrategy:
+    """Strategy for processing fan device data."""
+
+    @staticmethod
+    def process_data(
+        device_model: str, base_data: DreoGenericDeviceData, status: dict[str, Any]
+    ) -> DreoFanDeviceData:
+        """Process fan device specific data."""
+        # Initialize with required fields
+        fan_data = DreoFanDeviceData(
+            available=base_data["available"],
+            is_on=base_data["is_on"],
+        )
+
+        if "mode" in status:
+            fan_data["mode"] = str(status.get("mode", ""))
+
+        if "oscillate" in status:
+            fan_data["oscillate"] = bool(status.get("oscillate", False))
+
+        if "speed" in status and status.get("speed") is not None:
+            speed_range = (
+                FAN_DEVICE.get("config", {}).get(device_model, {}).get("speed_range")
+            )
+            if speed_range:
+                speed_value = float(status.get("speed", 0))
+                fan_data["speed_percentage"] = int(
+                    ranged_value_to_percentage(speed_range, speed_value)
+                )
+
+        return fan_data
+
+
+# Use generic device data type
+class DreoDataUpdateCoordinator(DataUpdateCoordinator[DreoDeviceData]):
+    """Class to manage fetching Dreo data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: HsCloud,
+        device_id: str,
+        model: str | None = None,
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.client = client
+        self.device_id = device_id
+        self.device_model = model
+        self.device_type = DEVICE_TYPE.get(model) if model else None
+        # Initialize with generic device data type
+        self.data = DreoGenericDeviceData(available=False, is_on=False)
+
+    async def _async_update_data(self) -> DreoDeviceData:
+        """Get device status from Dreo API and process it."""
+
+        # Create base data structure
+        base_data = DreoGenericDeviceData(available=False, is_on=False)
+
+        try:
+            # Get device status
+            status = await self.hass.async_add_executor_job(
+                self.client.get_status, self.device_id
+            )
+
+            if status is None:
+                return base_data
+
+            # Set common properties
+            base_data["available"] = status.get("connected", False)
+            base_data["is_on"] = status.get("power_switch", False)
```

Comment:
we could have a `GenericDataStrategy` class with its own `process_data()` method and move this to it.
then make `FanDataStrategy` inherit from it and make it call `super().process_data()`. that way this method gets simpler and the data extraction goes to the data classes